analytics/transform.go

```diff
@@ -191,22 +192,116 @@ func (event ParsedEvent) GetValue(field string) (interface{}, error) {
 		return nil, errors.New(fmt.Sprintf("Key %s not a valid atomic field", field))
 	}
 	if event[index] == "" {
-		return nil, errors.New(fmt.Sprintf("Field %s is empty", field))
+		return nil, errors.New(EmptyFieldErr)
 	}
 	kvPairs, err := enrichedEventFieldTypes[index].ParseFunction(enrichedEventFieldTypes[index].Key, event[index])
 	if err != nil {
 		return nil, err
 	}
-	if field == "contexts" || field == "derived_contexts" || field == "unstruct_event" {
-		// TODO: DRY HERE TOO?
-		output := make(map[string]interface{})
+
+	return kvPairs, nil
+}
+
+// GetValue returns the value for a provided atomic field, without processing the rest of the event.
+// For unstruct_event, it returns a map of only the data for the unstruct event.
+func (event ParsedEvent) GetValue(field string) (interface{}, error) {
+	kvPairs, err := event.getParsedValue(field)
+	if err != nil {
+		return nil, err
+	}
+	return kvPairs[0].Value, nil
+}
+
+// GetUnstructEventValue returns the value for a provided atomic field inside an event's unstruct_event field
+func (event ParsedEvent) GetUnstructEventValue(path ...interface{}) (interface{}, error) {
+	field := "unstruct_event"
+
+	kvPairs, err := event.getParsedValue(field)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the key/value pairs of the event field into a map
+	eventMap := make(map[string]interface{})
+	for _, pair := range kvPairs {
+		eventMap[pair.Key] = pair.Value
+	}
+
+	j, err := json.Marshal(eventMap)
+	if err != nil {
+		return nil, err
+	}
+
+	b := make([]interface{}, len(path))
+	for idx := range path {
+		b[idx] = path[idx]
+	}
+
+	el := json.Get(j, b...)
+	return el.GetInterface(), nil
+}
+
+// GetContextValue returns the value for a provided atomic field inside an event's contexts or derived_contexts
+func (event ParsedEvent) GetContextValue(contextName string, path ...interface{}) (interface{}, error) {
+	contextNames := []string{`contexts`, `derived_contexts`}
+	var contexts []interface{}
+	for _, c := range contextNames {
+		kvPairs, err := event.getParsedValue(c)
+		if err != nil && err.Error() != EmptyFieldErr {
+			return nil, err
+		}
+		// extract the key/value pairs of the event path into a map
+		eventMap := make(map[string]interface{})
 		for _, pair := range kvPairs {
-			output[pair.Key] = pair.Value
+			eventMap[pair.Key] = pair.Value
 		}
-		return output, nil
+		contexts = append(contexts, eventMap)
 	}
-	return kvPairs[0].Value, nil
 
+	var output []map[string]interface{}
```

Comment:
I _think_ the return type of this function should be a slice of interfaces, which may or may not be maps. If the path provided points to an int field, for example, I think it should return a map of ints.

I'm open to discussion on that though.

Reply:
@paulboocock I think this might relate to the conversation that happened here: https://github.com/snowplow/snowplow-golang-analytics-sdk/pull/14#discussion_r845953356, but I'm not sure I fully understand that whole conversation.

I wonder if the GetContextValue() function should return a slice of interface{} containing just the values it found for the given path, rather than a slice of maps.

So for example if the user searches for `mykey` in `contexts_acme_mycontext_1` rather than (pseudocode) `[contexts_acme_mycontext_1: {"mykey": "myvalue"}]` or `[{"mykey": "myvalue"}]`, it just returns `["myvalue"]`.

If the value at "mykey" is an object, we return a slice of maps, of course - point being it's an interface that doesn't include the key provided.

Does this conflict with your thinking? What do you think of this in general?

Reply:
@paulboocock just giving you a heads-up about this so you get a notification :D

Reply:
My apologies! 11 days to respond to a message is embarressing!

> it just returns ["myvalue"].

I really like what Colm proposes here, I think it simplifies the response and the logic in the function too.

The main use case of this function is simply to get the value at the path we've specified, so actually the response Colm suggests is an improvement, as you don't then need to figure out what the key is in the object to pick the value out of the response that you're actually trying to get to.

Reply:
Done!